rclrs/src/timer.rs

```diff
@@ -21,143 +30,303 @@ pub type TimerCallback = Box<dyn Fn(i64) + Send + Sync>;
+            unsafe { rcl_get_zero_initialized_timer() },
+        ));
+
+        unsafe {
+            let mut rcl_clock = clock.get_rcl_clock().lock().unwrap();
+            let mut rcl_context = context.rcl_context.lock().unwrap();
+
+            // SAFETY: Getting a default value is always safe.
+            let allocator = rcutils_get_default_allocator();
+
+            let _lifecycle = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
+            // SAFETY: We lock the lifecycle mutex since rcl_timer_init is not
+            // thread-safe.
+            rcl_timer_init(
+                &mut *rcl_timer.lock().unwrap(),
+                &mut *rcl_clock,
+                &mut *rcl_context,
+                period,
+                rcl_timer_callback,
+                allocator,
+            )
+        }
+        .ok()?;
+
+        let timer = Timer {
+            handle: TimerHandle { rcl_timer, clock },
+            callback: Arc::new(Mutex::new(Some(callback))),
+            last_elapse: Mutex::new(Duration::ZERO),
+            in_use_by_wait_set: Arc::new(AtomicBool::new(false)),
+        };
+        Ok(timer)
+    }
+
+    /// Force the timer to be called, even if it is not ready to be triggered yet.
+    /// We could consider making this public, but the behavior may confuse users.
+    fn call(&self) -> Result<(), RclrsError> {
+        // Keep track of the time elapsed since the last call. We need to run
+        // this before we trigger rcl_call.
+        let last_elapse = self.time_since_last_call().unwrap_or(Duration::ZERO);
+        *self.last_elapse.lock().unwrap() = last_elapse;
+
+        if let Err(err) = self.rcl_call() {
+            log_error!("timer", "Unable to call timer: {err:?}",);
+        }
+
+        let Some(callback) = self.callback.lock().unwrap().take() else {
+            log_error!(
+                "timer".once(),
+                "Timer is missing its callback information. This should not \
+                be possible, please report it to the maintainers of rclrs.",
+            );
+            return Ok(());
+        };
+
+        match callback {
+            AnyTimerCallback::Repeating(mut callback) => {
+                callback(self);
+                self.restore_callback(AnyTimerCallback::Repeating(callback));
+            }
+            AnyTimerCallback::OneShot(callback) => {
+                callback(self);
+                // We restore the callback as None because this was a
+                // one-shot which has been consumed.
+                self.restore_callback(AnyTimerCallback::None);
+            }
+            AnyTimerCallback::None => {
+                // Nothing to do here, just restore the callback.
+                self.restore_callback(AnyTimerCallback::None);
             }
         }
+
         Ok(())
     }
+
+    /// Updates the state of the rcl_timer to know that it has been called. This
+    /// should only be called by [`Self::execute`].
```

Comment:
Minor, this should be [`Self::call`] I think.

Reply:
Good catch, those functions were refactored after I wrote that documentation. Fixed here: [464d5d9](https://github.com/agalbachicar/ros2_rust/pull/5/commits/464d5d949016aee55a18bd78b6454332d12fb6f5)